refactor(gatsby-plugin-sharp): split single file into more maintainable chunks

wardpeet opened the pull request:
<!--
  Have any questions? Check out the contributing docs at https://gatsby.app/contribute, or
  ask in this Pull Request and a Gatsby maintainer will be happy to help :)
-->

## Description
This PR changes how the gatsby-plugin-sharp plugin handles images. Instead of processing our image queue right away we wait processing depending on your environment.

### Process
Inside queueImageResizing we are pushing every job into a memory queue and we don't care 
about processing here at all. Inside gatsby-node we're distinguishing develop and build.  

onPostBootstrap we're going to save our jobs into a cache. This cache is needed when we switch from develop to build mode. If no data changed we won't execute any queries which means no images are scheduled.

When we run `gatsby develop` we will use the `onCreateDevServer` hook to add an extra middelware to the express instance. Each file that doesn't exist yet gets fowarded to this route. We look into our queue if we know this url, if we do we process it and send the file over. If not we ignore and let express handle it.

`gatsby build` looks at our cached queue to see if it needs to do any processing. It loops over the queue and waits until all files are processed. We do this inside `onPostBuild` to make sure it's only ran in build mode.

## Updates
- Added jest-serializer-path to fix snapshots absolute paths
- Updated snapshots as sharps arguments got changed to reduce extension & toFormat checks
- Added integration tests to see if caching works between develop & build
- Added unit test to test image on demand.


## Related Issues
#10815
<!--
  Link to the issue that is fixed by this PR (if there is one)
  e.g. Fixes #1234, Addresses #1234, Related to #1234, etc.
-->

DSchau commented on packages/gatsby-plugin-sharp/src/gatsby-node.js:
```diff
@@ -1,10 +1,84 @@
+exports.onPreBuild = async ({ cache, boundActionCreators }, pluginOptions) => {
+  const cachedQueue = await getQueueFromCache(cache)
+
+  let promises = []
```
If nothing else--it's sorta confusing, so I'd maybe take another look at this!

wardpeet replied:
oh damn, yeah refactoring...

DSchau commented on packages/gatsby-plugin-sharp/src/gatsby-node.js:
```diff
@@ -1,10 +1,84 @@
+/**
+ * Build images on the fly when they are requested by the browser
+ */
+exports.onCreateDevServer = ({ app, cache }, pluginOptions) => {
```
This is slick!

KyleAMathews replied:
nifty!

Could we name this `onCreateDevelopmentServer`? I'm strongly allergic to shorten names for APIs :-)

Is the changes to develop.js not pushed yet? Plus the API docs changes?

pieh replied:
Oh, this is not new API (we have `onCreateDevServer` API for a while). So this is my bad for adding this API with shorten name :( mea culpa

wardpeet commented on packages/gatsby-plugin-sharp/.gitignore:
```diff
@@ -2,3 +2,6 @@
 /node_modules
 /gatsby-node.js
 /duotone.js
```
wasn't sure how to handle this so I just kept adding but good idea!

pieh commented on packages/gatsby-plugin-sharp/src/gatsby-node.js:
```diff
@@ -1,10 +1,84 @@
+/**
+ * save queue to the cache
+ */
+exports.onPostBootstrap = async ({ cache }) => {
```
in `develop` we can run queries after bootstrap (especially when working with files and editing them while `develop` is running) - it's good idea to wait for this event for initial save - but after that we would need to re-save queue

wardpeet replied:
Any suggestions? I Probably need to save it when dev server is created, just before [line 99](https://github.com/gatsbyjs/gatsby/pull/10964/files#diff-f60f2b266f94ca7cde8437d8451cfe39R66).

If queries change is there any mechanism to know what nodes are being removed because I probably want to remove unused images.

wardpeet commented on packages/gatsby-plugin-sharp/src/gatsby-node.js:
```diff
@@ -1,10 +1,93 @@
+ * Build images on the fly when they are requested by the browser
+ */
+exports.onCreateDevServer = async ({ app, cache, compiler }, pluginOptions) => {
+  compiler.hooks.done.tap(`gatsby-plugin-sharp`, () => {
```
In need to hook in on the compilation lifcycle to know whenever a new build is done.

wardpeet replied:
peer-review with @m-allanson 

> lets go for emitter.on('QUERY_QUEUE_DRAINED') instead of exposing compiler

wardpeet commented:
So what I haven't fixed yet is removing images that should not be processed because a page or query changed.

wardpeet commented on packages/gatsby-plugin-sharp/src/gatsby-node.js:
```diff
@@ -1,10 +1,93 @@
+    queue.delete(req.originalUrl)
+
+    // JSON.stringify doesn't work on an Map so we need to convert it to an array
+    cache.set(`queue`, Array.from(queue))
```
I should use ` saveQueueToCache `

wardpeet commented:
should be good to go

pieh commented on packages/gatsby-plugin-sharp/src/index.js:
```diff
@@ -464,13 +176,29 @@ function queueImageResizing({ file, args = {}, reporter }) {
     aspectRatio,
-    finishedPromise,
+    // finishedPromise is needed to not break our API (https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-sqip/src/extend-node-type.js#L115)
+    finishedPromise: () => {
```
Will this actually work? Maybe we should make lazy image generation more opt-in? So `gatsby-transformer-sharp` and `gatsby-remark-images` specifically opt-in into it with something like extra `lazy` flag? and keep current behaviour of scheduling job asap if caller didn't opt into it?

wardpeet replied:
I'm checking if I don't break but I actually like this approach as no one needs to opt-in for speed. Most plugins have enough with just queuing a job.

I probably write an  RFC about cleaning this plugin up, it will need some changes to core to expose reporter and createJob from a require.

wardpeet replied:
@pieh I added a test to demonstrate that this is working. I also tested it with the examples/using-sqip where this is used and it is working

wardpeet commented:
> But there is also one unsolved issue:
some images can be scheduled to be processed, but later on should be invalidated that currently doesn't happen - for most part this would just mean extra unneded work, but there is edge cases where this would cause straight errors - for example image was scheduled, but later on source image was deleted and now this scheduled task have no way of finishing and we would actually throw errors about trying to read not-existing files? Initially we can just add some sanity checks to ensure input file exists before trying to process it I think, but later on I think we need to keep track of queries to remove stale tasks	

Is it ok if I do file test in this PR and do the query stuff in a follow up PR.

wardpeet commented on packages/gatsby-plugin-sharp/src/processFile.js:
```diff
@@ -0,0 +1,182 @@
+
+const writeFileAsync = (file, buffer) =>
+  new Promise((resovle, reject) => {
+    fs.writeFile(file, buffer, err => {
```
sweet i'll inline, it's better that way :)

wardpeet commented on packages/gatsby-plugin-sharp/src/index.js:
```diff
@@ -464,13 +176,29 @@ function queueImageResizing({ file, args = {}, reporter }) {
 
+  // Create job and add it to the queue, the queue will be processed inside gatsby-node.js
+  const job = {
+    file,
```
good point, I wasn't thinking of this, i'll normalize it.

wardpeet replied:
I can even remove it as only absolutePath is needed from file (if not mistaken) need to check to be sure

wardpeet commented on packages/gatsby-plugin-sharp/src/processFile.js:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    if (
+      (file.extension === `webp` && args.toFormat === ``) ||
```
yes 🤦‍♂️

pieh commented:
> Is it ok if I do file test in this PR and do the query stuff in a follow up PR.

👍 from me - it will be much easier to review

wardpeet commented:
Some updates:
- Added jest-serializer-path to fix snapshots absolute paths
- Updated snapshots as sharps arguments got changed to reduce extension & toFormat checks
- Added integration tests to see if caching works between develop & build
- Added unit test to test image on demand.

I'm pretty confident of shipping this when all tests are green. I've tested in on www and some examples and I did not run into any issues. Big issues take some time to get resized but aren't blocking for the user 👍

wardpeet commented:
when this get approved, let me know so I can merge it myself and write a proper commit message :)

jlengstorf commented:
This is so exciting!

pieh commented:
@m-allanson yes! I hope to do this today

jlengstorf commented:
We need a shipit squirrel emoji reaction.

pieh commented on packages/gatsby-plugin-sharp/src/gatsby-node.js:
```diff
@@ -1,10 +1,96 @@
+  }
+
+  // JSON.stringify doesn't work on an Map so we need to convert it to an array
+  return cache.set(`queue`, Array.from(queue))
```
This is only thing that could potentially be blocker.

Just to get it out of the way - I suggested using `cache` initially, so this is my fault.

Since we started looking into cache issues, I started paying more attention to potential problems with it. Problem with using `cache` instead of redux state here would be that we potentially can have inconsistent state. 

Processing queue is linked to query running (not exclusively but major usage of it is). State that triggers query running is stored in redux state (there are other triggers, but things we persist is stored there). Scenario when processing queue fails to persist (to `cache`), but redux state persist correctly can lead to situations when images that are supposed to be created during production build, aren't, and build completes successfully (no errors thrown).

Long story short - we can't store related chunks of data in different persistence engines. We need to use redux state here. 

Luckily we have easy way to do that through [setPluginStatus](https://www.gatsbyjs.org/docs/actions/#setPluginStatus) action:
```
exports.whateverAPI = async ({ actions }) => {
  actions.setPluginStatus({ queue: Array.from(queue) })
}
``` 

Reading it is not as nice, but still OK-ish:
```js
let getPluginState = null
exports.onPreBootstrap = ({ store }) => {
  getPluginState = () => {
    const pluginStatus = store.getState().status.plugins[`gatsby-plugin-sharp`]
    if (pluginStatus && pluginStatus.queue) {
      return new Map(pluginStatus.queue)
    }
    return new Map()
  }
}
```
(ideally we add helper function for this to gatsby, but I would suggest using this to be backward compatible as there aren't any breaking changes here)

So all this can potentially increase problems with persisting redux state (higher memory usage / long writes), which is not great :/ Some issues for reference - https://github.com/gatsbyjs/gatsby/issues/8125 https://github.com/gatsbyjs/gatsby/issues/6611 https://github.com/gatsbyjs/gatsby/issues/6656

So I hate to even suggest this given how great feature this is - but should we give a way to opt-out of it? (if other folks agree with my essay that is - I would very happy to be proven wrong about inconsistency issue - it happened before)

wardpeet replied:
@pieh thanks a lot! I'll see what I can do with it. I'll add a new pluginOption `lazyImageGenaration` (if you have a better please tell me). Which is set by default on true.

It's sad to hear that redux state persisting might cause memory issues. We maybe should try to move to something else or be smarter about our store, like don't load it upfront unless it's asked for and clear objects when they aren't accessed anymore

wardpeet replied:
I did the changes, unsure if this is what you meant 😛 @pieh

pieh replied:
Memory issues mostly are related to saving state to file - and not saving part but serializing ( https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/redux/index.js#L91 ).

Just the way `stringify` (either `JSON.stringify` or one from `json-stringify-safe`) works, is that it creates many intermediatte strings which add up in memory and not being garbage collected quick enough by Node's V8 engine.

We tried some streaming alternatives but this wasn't really working for us (at least our implementation of using it). See https://github.com/gatsbyjs/gatsby/pull/9896 for revert of it and linked comment in PR description that summarise issue we had with it.

wardpeet commented on www/README.md:
````diff
@@ -39,15 +39,3 @@ GATSBY_FEEDBACK_KEY_STARTERLIB=ADD_KEY
-GATSBY_SCREENSHOT_PLACEHOLDER=true
-```
-
-For more information checkout [`gatsby-transformer-screenshot` docs](http://www.gatsbyjs.org/packages/gatsby-transformer-screenshot#placeholder-image).
````
sweet i'll revert

wardpeet commented on packages/gatsby-plugin-sharp/src/index.js:
```diff
@@ -464,19 +185,37 @@ function queueImageResizing({ file, args = {}, reporter }) {
-    finishedPromise,
+    // finishedPromise is needed to not break our API (https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-sqip/src/extend-node-type.js#L115)
+    finishedPromise: {
+      then: (resolve, reject) => {
```
use new Promise where you can :) I once read that this worked with async/await just unsure where I found it, tried to google to put a reference here but 🤷‍♂️

wardpeet commented on packages/gatsby-transformer-toml/src/__tests__/__snapshots__/gatsby-node.js.snap:
```diff
@@ -81,7 +81,7 @@ Array [
     test_array2 = [ \\"Test #11 ]proved that\\", \\"Experiment #9 was a success\\" ]
     another_test_string = \\" Same thing, but with a string #\\"
-    harder_test_string = \\" And when \\\\\\"'s are in the string, along with # \\\\\\"\\"
+    harder_test_string = \\" And when /\\"'s are in the string, along with # /\\"\\"
```
yeah it is 🤔 it's probably a bug as they shouldn't touch it but it does it consistently so it's not breaking. I'll make a bug on their repo and perhaps fix it

pieh replied:
I don't mind if this is consistent (and unit tests pass on both circleCI and azure so we should be fine). Was just curious where this change came from

KyleAMathews commented:
I don't think this will add that much more data to the cache right? Seems minor relative to the benefits we'll get.

We need to figure out more scalable ways of storing the cache but given that this is a relatively small % of the cache data, it won't cause more trouble to people already having trouble.

pieh commented:
for gatsbyjs.org on cold cache it will add ~12k entries with shape like this https://github.com/gatsbyjs/gatsby/blob/03d4e7cbd2dbdf3adf576c202c39ef8f9b024a3d/packages/gatsby-plugin-sharp/src/__tests__/__snapshots__/index.js.snap#L78-L95

I'm doing actual testing right now - yesterday I was doing code review only. So will report `redux-state.json` before/after comparison.

wardpeet commented:
@DSchau thanks! I hope we can get more of these integration tests running to test some core pieces.

@pieh if it helps I can omit default values from the args to make the cache smaller. I could do the same for inputPath & outputPath (strip the program root, where possible). I added an escape hatch called `lazyImageGeneration` that you can set to false if we want to opt out. I should add it to the README.md

pieh commented:
>@pieh if it helps I can omit default values from the args to make the cache smaller.

If it's relatively straight forward to do - let's strip as much as we can. But with changes you did to not persist entire File node earlier we are pretty good already (so this is nice to have, not a blocker)

I'm still checking and validating this, but initial checks redux-state.json for gatsbyjs.org (with screenshot placeholders - so not trully production build, but still ~2400 image thumbnails) increased from `26314316` to `30325343` bytes so ~15% (that's after running gatsby develop and not visiting site, so all images processing tasks are in cache).

Moocar commented:
@wardpeet Thanks for the heads up on this PR. I'm working on making the graphql resolvers run in other processes. To do this, those resolvers can't rely on accessing the redux store directly, since it runs in another process. This was fine, except that `plugin-sharp` manually requires the redux store in order to call the `setJob` bound action creators. When running in the main process, that's fine, but when it's running in another process, it actually ends up creating a redux store per process, which means the setJob calls are entirely uncoordinated.

I've been teasing that out so that the job API functions are provided to plugin-sharp instead of it having to require them. This PR makes it even harder to run plugin-sharp in another process because it introduces more globals and dependence on the gatbsy-node APIs. But the benefits are obviously massive so I recommend merging this PR, and I'll have a think about whether it's possible to make this run in parallel resolvers land.

Moocar commented:
I think it might be possible to make parallel resolvers work here. Calls to `fluid`, `fixed` etc don't cause side effects (apart from reading image size from disk and generating base64) until they add the processing job to the queue. So theoretically, we could change `queueImageResizing` so that if it's running in another process, it instead makes a call to the main process where it puts the job into the in memory queue. Otherwise it runs as normal.

I'll wait till this is merged, and then have more of a think about it.

wardpeet commented:
@pieh I could run this to remove default ones:
https://repl.it/@wardpeet/omit-defaults

I could even remove lodash if needed. WDYT? should I sneak it in and than we're good to go?

wardpeet commented:
@pieh found an issue where we could remove images from disk between develop & build. At the moment we fail hard 😄. The idea is to do a simple file_exists on all files that are inside the cached version and ignore those errors. We will still fail hard if our current queue has errors.

Any issues with this approach? I would love to get this merged 😄

pieh commented:
I was thinking yesterday and today about this and this approach also have problems:
- use default starter
- run gatsby develop (don't open browser) - image in https://github.com/gatsbyjs/gatsby-starter-default/blob/master/src/components/image.js won't be processed but job will be cached
- delete "src/images/gatsby-astronaut.png"

With silencing error build succeed, but we have missing assets for that image. I think build should fail.

I don't have good approach yet. We might need for gatsby to expose more information to `gatsby-plugin-sharp` to properly invalidate stale jobs

wardpeet commented:
Turned lazy loading off so we can merge this without having impact 👍 

cc @pieh

sidharthachatterjee commented:
Published in gatsby-plugin-sharp@2.0.26

KyleAMathews commented:
Just a thought — if image processing jobs are marked as a dependency of a query — then when we re-run the query, we would delete all the old jobs associated with it (and recreate them of course when the query runs). This would solve I believe the stale job problem because deleting a file between dev and build would dirty the query making it re-run.

muescha commented:
any plans to turn the `lazyImageGeneration` on?

wesbos commented:
Please turn this on :) My build takes ~8 mins on a really fast macbook pro before I can even see my site.
